internal/pipeline/types_test.go

```diff
@@ -0,0 +1,258 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package pipeline_test
+
+import (
+	"reflect"
+	"testing"
+
+	. "github.com/GoogleContainerTools/kpt/internal/pipeline"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+func parseFromString(input string) (Pipeline, error) {
+	var p Pipeline
+	err := yaml.Unmarshal([]byte(input), &p)
+	if err != nil {
+		return p, err
+	}
+	return p, nil
+}
+
+func isFunctionEqual(t *testing.T, f1, f2 Function) bool {
+	if reflect.DeepEqual(f1.Config, f2.Config) {
+		return reflect.DeepEqual(f1, f2)
+	}
+	// Config objects cannot be compared directly
+	f1ConfigString, err := yaml.String(f1.Config.Node)
+	assert.NoError(t, err)
+	f2ConfigString, err := yaml.String(f2.Config.Node)
+	assert.NoError(t, err)
+
+	// Compare the functions regardless of Config
+	f1ConfigCopy := f1.Config
+	f2ConfigCopy := f2.Config
+	f1.Config = NewFunctionConfig(nil)
+	f2.Config = NewFunctionConfig(nil)
+	result := assert.EqualValues(t, f1, f2) &&
+		assert.EqualValues(t, f1ConfigString, f2ConfigString)
+	f1.Config = f1ConfigCopy
+	f2.Config = f2ConfigCopy
+	return result
+}
+
+func isFunctionSliceEqual(t *testing.T, fs1, fs2 []Function) bool {
+	if len(fs1) != len(fs2) {
+		return false
+	}
+	for i := range fs1 {
+		if !isFunctionEqual(t, fs1[i], fs2[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func isPipelineEqual(t *testing.T, p1, p2 Pipeline) bool {
+	p1Transformers := p1.Transformers
+	p1.Transformers = []Function{}
+	p2Transformers := p2.Transformers
+	p2.Transformers = []Function{}
+	if !isFunctionSliceEqual(t, p1Transformers, p2Transformers) {
+		return false
+	}
+
+	p1Generators := p1.Generators
+	p1.Generators = []Function{}
+	p2Generators := p2.Generators
+	p2.Generators = []Function{}
+	if !isFunctionSliceEqual(t, p1Generators, p2Generators) {
+		return false
+	}
+
+	p1Validators := p1.Validators
+	p1.Validators = []Function{}
+	p2Validators := p2.Validators
+	p2.Validators = []Function{}
+	if !isFunctionSliceEqual(t, p1Validators, p2Validators) {
+		return false
+	}
+
+	result := assert.EqualValues(t, p1, p2)
+	p1.Transformers = p1Transformers
+	p1.Generators = p1Generators
+	p1.Validators = p1Validators
+	p2.Transformers = p2Transformers
+	p2.Generators = p2Generators
+	p2.Validators = p2Validators
+	return result
+}
+
+func TestPipelineParse(t *testing.T) {
+	type testcase struct {
+		Input    string
+		Expected Pipeline
+		Error    bool
+	}
+
+	testcases := []testcase{
+		{
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
```

Comment:
Other interesting corner cases here:

1. sources omitted and defaulting to `./*`
2. `sources: []`
3.  transformers omitted and default to `[]`
4. `transformers: []`

Reply:
This PR is only for defining the types. The default values will be tested with pipeline read/write.